src/energy.h

```diff
@@ -1571,12 +1574,70 @@ class SASAEnergy : public Energybase {
      * @param cosolute_concentration in particles per angstrom cubed
      * @param probe_radius in angstrom
      */
-    SASAEnergy(const Space& spc, double cosolute_concentration = 0.0, double probe_radius = 1.4);
-    SASAEnergy(const json& j, const Space& spc);
+    FreeSASAEnergy(const Space& spc, double cosolute_concentration = 0.0, double probe_radius = 1.4);
+    FreeSASAEnergy(const json& j, const Space& spc);
     double energy(Change& change) override;
+    const std::vector<double>& getAreas() const { return sasa; }
 }; //!< SASA energy from transfer free energies
 #endif
 
+class SASAEnergyBase : public Energybase {
+
+  protected:
+    std::vector<double> areas; //!< Target buffer for calculated surface areas
+    std::vector<double> radii; //!< Radii buffer for all particles
+    Space& spc;
+    double cosolute_concentration; //!< co-solute concentration (mol/l)
+    SASA sasa;                     //!< performs neighbour searching and subsequent sasa calculation
+    Average<double> mean_surface_area;
+
+  private:
+    void to_json(json& j) const override;
+    void sync(Energybase* energybase_ptr, const Change& change) override;
+    void init() override;
+
+  public:
+    /**
+     * @param spc
+     * @param cosolute_concentration in particles per angstrom cubed
+     * @param probe_radius in angstrom
+     * @param n_slices_per_atom number of slices of spheres in SASA calculation
+     */
+    SASAEnergyBase(Space& spc, double cosolute_concentration = 0.0, double probe_radius = 1.4,
+                   int n_slices_per_atom = 20);
+    SASAEnergyBase(const json& j, Space& spc);
+    const std::vector<double>& getAreas() const { return areas; }
+    double energy(Change& change) override;
+}; //!< SASA energy from transfer free energies with SASA calculation each step
+
+class SASAEnergy : public SASAEnergyBase {
+
+  private:
+    std::vector<std::vector<int>>
```

Comment:
Perhaps not important, but vector of vector is fragmented in memory. An Eigen matrix or our custom matrix class could be used instead, see `src/aux/multimatrix.h`